File: packaging/mac/KeikoStockAI.m
#import <Cocoa/Cocoa.h>
#import <WebKit/WebKit.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

@interface AppDelegate : NSObject <NSApplicationDelegate>
@property(nonatomic, strong) NSWindow *window;
@property(nonatomic, strong) WKWebView *webView;
@property(nonatomic, strong) NSTask *serverTask;
@property(nonatomic, assign) NSInteger port;
@end

@implementation AppDelegate

- (void)applicationDidFinishLaunching:(NSNotification *)notification {
    self.port = 8123;
    [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
    [self startBackend];
    [self createWindow];
    [self waitForBackend:0];
}

- (void)applicationWillTerminate:(NSNotification *)notification {
    if (self.serverTask && self.serverTask.isRunning) {
        [self.serverTask terminate];
    }
}

- (BOOL)applicationShouldHandleReopen:(NSApplication *)sender hasVisibleWindows:(BOOL)flag {
    if (!flag) {
        if (!self.window) {
            [self createWindow];
            [self waitForBackend:0];
        } else {
            [self.window makeKeyAndOrderFront:nil];
        }
        [NSApp activateIgnoringOtherApps:YES];
    }
    return YES;
}

- (void)createWindow {
    WKWebViewConfiguration *configuration = [[WKWebViewConfiguration alloc] init];
    configuration.websiteDataStore = [WKWebsiteDataStore defaultDataStore];
    self.webView = [[WKWebView alloc] initWithFrame:NSZeroRect configuration:configuration];

    self.window = [[NSWindow alloc]
        initWithContentRect:NSMakeRect(0, 0, 1220, 860)
                  styleMask:(NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable | NSWindowStyleMaskResizable)
                    backing:NSBackingStoreBuffered
                      defer:NO];
    self.window.title = @"Keiko Stock AI";
    self.window.minSize = NSMakeSize(1040, 720);
    self.window.contentView = self.webView;
    [self.window center];
    [self.window makeKeyAndOrderFront:nil];
    [NSApp activateIgnoringOtherApps:YES];
}

- (void)startBackend {
    if ([self backendIsHealthy]) {
        return;
    }
    self.port = [self availablePortStartingAt:self.port];

    NSString *resourcesPath = [[NSBundle mainBundle] resourcePath];
    NSString *appRoot = [resourcesPath stringByAppendingPathComponent:@"app"];

    NSFileManager *fm = [NSFileManager defaultManager];
    NSURL *supportURL = [[fm URLsForDirectory:NSApplicationSupportDirectory inDomains:NSUserDomainMask].firstObject
        URLByAppendingPathComponent:@"Keiko Stock AI" isDirectory:YES];
    NSURL *dataURL = [supportURL URLByAppendingPathComponent:@"data" isDirectory:YES];
    NSURL *logURL = [supportURL URLByAppendingPathComponent:@"logs" isDirectory:YES];
    [fm createDirectoryAtURL:dataURL withIntermediateDirectories:YES attributes:nil error:nil];
    [fm createDirectoryAtURL:logURL withIntermediateDirectories:YES attributes:nil error:nil];

    NSString *serverLog = [[logURL path] stringByAppendingPathComponent:@"server.log"];
    [fm createFileAtPath:serverLog contents:nil attributes:nil];
    NSFileHandle *logHandle = [NSFileHandle fileHandleForWritingAtPath:serverLog];
    [logHandle seekToEndOfFile];

    NSTask *task = [[NSTask alloc] init];
    task.launchPath = [self pythonExecutable];
    task.currentDirectoryPath = appRoot;
    task.arguments = @[
        @"-B", @"-m", @"uvicorn", @"backend.app:app",
        @"--host", @"127.0.0.1",
        @"--port", [NSString stringWithFormat:@"%ld", (long)self.port]
    ];

    NSMutableDictionary *environment = [[[NSProcessInfo processInfo] environment] mutableCopy];
    environment[@"PYTHONPATH"] = appRoot;
    environment[@"KEIKO_DATA_DIR"] = [dataURL path];
    environment[@"PATH"] = @"/opt/homebrew/Caskroom/miniconda/base/bin:/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin";
    task.environment = environment;
    task.standardOutput = logHandle;
    task.standardError = logHandle;

    @try {
        [task launch];
        self.serverTask = task;
    } @catch (NSException *exception) {
        [self showStartupError:[NSString stringWithFormat:@"后端启动失败：%@\n日志：%@", exception.reason, serverLog]];
    }
}

- (BOOL)backendIsHealthy {
    NSString *healthURL = [NSString stringWithFormat:@"http://127.0.0.1:%ld/api/health", (long)self.port];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:healthURL]];
    request.timeoutInterval = 0.35;

    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
    __block BOOL healthy = NO;
    NSURLSessionDataTask *task = [[NSURLSession sharedSession]
        dataTaskWithRequest:request
          completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
              NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];
              healthy = (statusCode == 200);
              dispatch_semaphore_signal(semaphore);
          }];
    [task resume];
    dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.45 * NSEC_PER_SEC)));
    [task cancel];
    return healthy;
}

- (NSInteger)availablePortStartingAt:(NSInteger)startPort {
    for (NSInteger candidate = startPort; candidate < startPort + 40; candidate++) {
        int socketFD = socket(AF_INET, SOCK_STREAM, 0);
        if (socketFD < 0) {
            continue;
        }

        int opt = 1;
        setsockopt(socketFD, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

        struct sockaddr_in address;
        memset(&address, 0, sizeof(address));
        address.sin_family = AF_INET;
        address.sin_port = htons((uint16_t)candidate);
        inet_pton(AF_INET, "127.0.0.1", &address.sin_addr);

        int result = bind(socketFD, (struct sockaddr *)&address, sizeof(address));
        close(socketFD);
        if (result == 0) {
            return candidate;
        }
    }
    return startPort + 100;
}

- (void)waitForBackend:(NSInteger)attempt {
    if (attempt >= 80) {
        [self showStartupError:@"后端启动超时。请确认 Python 已安装依赖：pip install -r requirements.txt"];
        return;
    }

    NSString *healthURL = [NSString stringWithFormat:@"http://127.0.0.1:%ld/api/health", (long)self.port];
    NSURLSessionDataTask *task = [[NSURLSession sharedSession]
        dataTaskWithURL:[NSURL URLWithString:healthURL]
      completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
          NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];
          dispatch_async(dispatch_get_main_queue(), ^{
              if (statusCode == 200) {
                  NSString *appURL = [NSString stringWithFormat:@"http://127.0.0.1:%ld/", (long)self.port];
                  [self.webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:appURL]]];
              } else {
                  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.25 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                      [self waitForBackend:(attempt + 1)];
                  });
              }
          });
      }];
    [task resume];
}

- (NSString *)pythonExecutable {
    NSArray<NSString *> *candidates = @[
        @"/opt/homebrew/Caskroom/miniconda/base/bin/python3",
        @"/opt/homebrew/bin/python3",
        @"/usr/local/bin/python3",
        @"/usr/bin/python3"
    ];
    NSFileManager *fm = [NSFileManager defaultManager];
    for (NSString *candidate in candidates) {
        if ([fm isExecutableFileAtPath:candidate]) {
            return candidate;
        }
    }
    return @"/usr/bin/python3";
}

- (void)showStartupError:(NSString *)message {
    NSAlert *alert = [[NSAlert alloc] init];
    alert.alertStyle = NSAlertStyleCritical;
    alert.messageText = @"Keiko Stock AI 启动失败";
    alert.informativeText = message;
    [alert runModal];
}

@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSApplication *app = [NSApplication sharedApplication];
        AppDelegate *delegate = [[AppDelegate alloc] init];
        app.delegate = delegate;
        [app run];
    }
    return 0;
}

File: packaging/mac/KeikoStockAI.swift
import Cocoa
import Foundation
import WebKit

final class AppDelegate: NSObject, NSApplicationDelegate {
    private var window: NSWindow?
    private var webView: WKWebView?
    private var serverProcess: Process?
    private let port = 8123

    func applicationDidFinishLaunching(_ notification: Notification) {
        NSApp.setActivationPolicy(.regular)
        startBackend()
        createWindow()
        waitForBackend(attempt: 0)
    }

    func applicationWillTerminate(_ notification: Notification) {
        serverProcess?.terminate()
    }

    private func createWindow() {
        let configuration = WKWebViewConfiguration()
        configuration.websiteDataStore = .default()
        let webView = WKWebView(frame: .zero, configuration: configuration)
        self.webView = webView

        let window = NSWindow(
            contentRect: NSRect(x: 0, y: 0, width: 1220, height: 860),
            styleMask: [.titled, .closable, .miniaturizable, .resizable],
            backing: .buffered,
            defer: false
        )
        window.title = "Keiko Stock AI"
        window.minSize = NSSize(width: 1040, height: 720)
        window.contentView = webView
        window.center()
        window.makeKeyAndOrderFront(nil)
        NSApp.activate(ignoringOtherApps: true)
        self.window = window
    }

    private func startBackend() {
        guard let appRoot = Bundle.main.resourceURL?.appendingPathComponent("app") else {
            showStartupError("找不到打包资源目录。")
            return
        }

        let supportDir = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/Keiko Stock AI", isDirectory: true)
        let dataDir = supportDir.appendingPathComponent("data", isDirectory: true)
        let logDir = supportDir.appendingPathComponent("logs", isDirectory: true)
        try? FileManager.default.createDirectory(at: dataDir, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: logDir, withIntermediateDirectories: true)

        let process = Process()
        process.executableURL = pythonExecutable()
        process.currentDirectoryURL = appRoot
        process.arguments = [
            "-B", "-m", "uvicorn", "backend.app:app",
            "--host", "127.0.0.1",
            "--port", "\(port)"
        ]

        var environment = ProcessInfo.processInfo.environment
        environment["PYTHONPATH"] = appRoot.path
        environment["KEIKO_DATA_DIR"] = dataDir.path
        environment["PATH"] = [
            "/opt/homebrew/Caskroom/miniconda/base/bin",
            "/opt/homebrew/bin",
            "/usr/local/bin",
            "/usr/bin",
            "/bin",
            "/usr/sbin",
            "/sbin"
        ].joined(separator: ":")
        process.environment = environment

        let logURL = logDir.appendingPathComponent("server.log")
        FileManager.default.createFile(atPath: logURL.path, contents: nil)
        if let handle = try? FileHandle(forWritingTo: logURL) {
            handle.seekToEndOfFile()
            process.standardOutput = handle
            process.standardError = handle
        }

        do {
            try process.run()
            serverProcess = process
        } catch {
            showStartupError("后端启动失败：\(error.localizedDescription)\n日志：\(logURL.path)")
        }
    }

    private func waitForBackend(attempt: Int) {
        guard attempt < 80 else {
            showStartupError("后端启动超时。请确认 Python 已安装依赖：pip install -r requirements.txt")
            return
        }

        let url = URL(string: "http://127.0.0.1:\(port)/api/health")!
        URLSession.shared.dataTask(with: url) { _, response, _ in
            let ok = (response as? HTTPURLResponse)?.statusCode == 200
            DispatchQueue.main.async {
                if ok {
                    self.webView?.load(URLRequest(url: URL(string: "http://127.0.0.1:\(self.port)/")!))
                } else {
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.25) {
                        self.waitForBackend(attempt: attempt + 1)
                    }
                }
            }
        }.resume()
    }

    private func pythonExecutable() -> URL {
        let candidates = [
            "/opt/homebrew/Caskroom/miniconda/base/bin/python3",
            "/opt/homebrew/bin/python3",
            "/usr/local/bin/python3",
            "/usr/bin/python3"
        ]
        for candidate in candidates where FileManager.default.isExecutableFile(atPath: candidate) {
            return URL(fileURLWithPath: candidate)
        }
        return URL(fileURLWithPath: "/usr/bin/python3")
    }

    private func showStartupError(_ message: String) {
        let alert = NSAlert()
        alert.alertStyle = .critical
        alert.messageText = "Keiko Stock AI 启动失败"
        alert.informativeText = message
        alert.runModal()
    }
}

let app = NSApplication.shared
let delegate = AppDelegate()
app.delegate = delegate
app.run()
